File: Top_100_Liked_Questions/152._Maximum_Product_Subarray/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ans = mn = mx = nums[0]
        for i in range(1, len(nums)):
            n = nums[i]
            mx, mn = max(n, mn*n, mx*n), min(n, mn*n, mx*n)
            ans = max(ans, mx)
        
        return ans
    
    
    def maxProduct2(self, nums: List[int]) -> int:
        # brute force
        ans = float('-inf')
        for i in range(len(nums)):
            curr = 1
            for j in range(i, len(nums)):
                curr *= nums[j]
                ans = max(ans, curr)
        
        return ans

File: Top_100_Liked_Questions/152._Maximum_Product_Subarray/solution.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums.length == 0)
            return 0;

        int max = nums[0];
        int min = nums[0];
        int ans = nums[0];

        for (int i=1; i<nums.length; i++) {
            int curr = nums[i];
            int temp_max = Math.max(Math.max(curr, max * curr), min * curr);
            min = Math.min(Math.min(curr, max * curr), min * curr);
            max = temp_max;

            ans = Math.max(ans, max);
        }

        return ans;
    }
}